alexandria/core/visibilities.py

```diff
@@ -0,0 +1,119 @@
+import inspect
+
+from django.conf import settings
+from django.contrib.auth.models import AnonymousUser
+from django.core.exceptions import ImproperlyConfigured
+
+from . import models
+from .collections import list_duplicates
+
+
+def filter_queryset_for(model):
+    """Decorate function to define filtering of queryset of specific model."""
+
+    def decorate(fn):
+        if not hasattr(fn, "_visibilities"):
+            fn._visibilities = []
+
+        fn._visibilities.append(model)
+        return fn
+
+    return decorate
+
+
+class BaseVisibility(object):
+    """Basic visibility classes to be extended by any visibility implementation.
+
+    In combination with the decorator `@filter_queryset_for` a custom visibility class
+    can define filtering on basis of models.
+
+    A custom visibility class could look like this:
+    ```
+    >>> from alexandria.core.visibilities import BaseVisibility
+    ... from alexandria.core.models import BaseModel, Document, File
+    ...
+    ...
+    ... class CustomVisibility(BaseVisibility):
+    ...     @filter_queryset_for(BaseModel)
+    ...     def filter_queryset_for_all(self, queryset, request):
+    ...         return queryset.filter(created_by_user=request.user.username)
+    ...
+    ...     @filter_queryset_for(Document)
+    ...     def filter_queryset_for_document(self, queryset, request):
+    ...         return queryset.exclude(category__slug='protected-category')
+    ...
+    ...     @filter_queryset_for(File)
+    ...     def filter_queryset_for_file(self, queryset, request):
+    ...         # Limitations for `Document` should also be enforced on `File`.
+    ...         return queryset.exclude(document__category__slug='protected-category')
+    """
+
+    def __init__(self):
+        queryset_fns = inspect.getmembers(self, lambda m: hasattr(m, "_visibilities"))
+        queryset_nodes = [
+            node.__name__ for _, fn in queryset_fns for node in fn._visibilities
+        ]
+        queryset_nodes_dups = list_duplicates(queryset_nodes)
+        if queryset_nodes_dups:
+            raise ImproperlyConfigured(
+                f"`filter_queryset_for` defined multiple times for "
+                f"{', '.join(queryset_nodes_dups)} in {str(self)}"
+            )
+        self._filter_querysets_for = {
+            node: fn for _, fn in queryset_fns for node in fn._visibilities
+        }
+
+    def filter_queryset(self, model, queryset, request):
+        for cls in model.mro():
+            if cls in self._filter_querysets_for:
+                return self._filter_querysets_for[cls](queryset, request)
+
+        return queryset
+
+
+class Any(BaseVisibility):
+    """No restrictions, all models are exposed."""
+
+    pass
+
+
+class OwnAndAdmin(BaseVisibility):
```

Comment:
Might need a docstring for a quick explanation

Reply:
Done